fix building error for `error: fallthrough annotation in unreachable code`

shushanhf opened the pull request:
```
/home/qiao/work_qiao/runtime/src/coreclr/interop/comwrappers.cpp:740:21: error: fallthrough annotation in unreachable code [-Werror,-Wimplicit-fallthrough]
                    FALLTHROUGH;
                    ^
/home/qiao/work_qiao/runtime/src/native/minipal/utils.h:25:25: note: expanded from macro 'FALLTHROUGH'
#    define FALLTHROUGH [[fallthrough]]
```

shushanhf commented:
Hi, @jkotas 
Who can help reviewing this PR?
Thanks

AaronRobinsonMSFT commented:
Is this on a `Release` or `Debug`/`Checked` build?

AaronRobinsonMSFT commented:
The problem here and something the above code now doesn't handle is this code is calling into user defined [`ICustomQueryInterface.GetInterface()`](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.icustomqueryinterface.getinterface#system-runtime-interopservices-icustomqueryinterface-getinterface(system-guid@-system-intptr@)). The user could cast any `int` to `CustomQueryInterfaceResult` and then the above logic is flawed.

am11 commented on src/coreclr/interop/comwrappers.cpp:
```diff
@@ -752,6 +749,8 @@ HRESULT ManagedObjectWrapper::QueryInterface(
                     // the same way that would occur if the managed object lacked
                     // an ICustomQueryInterface implementation.
                     break;
+                default:
```
> What tool is complaing about this?

Was wondering the same, I don't get error with clang 16, 17 and 18, at least not with cross-compile https://godbolt.org/z/41fzaMcGn

shushanhf replied:
> This is a subtle behavior change if a new value is added to `TryInvokeICustomQueryInterfaceResult`.

Yes, the first patch of this PR  just resolved the compiling error.
It is better to support reporting error within the release mode when passing a unexpected value.

shushanhf replied:
> > What tool is complaing about this?
> 
> Was wondering the same, I don't get error with clang 16, 17 and 18, at least not with cross-compile https://godbolt.org/z/41fzaMcGn

On Debug build on the LoongArch64-linux with clang-13 and arm64 linux with clang-8.
I tested it on AMD64-linux that is also failed when building the coreclr.
I met this error every time with native linux compile not cross-compile.

AaronRobinsonMSFT replied:
I don't really understand what we are missing here. There is no unreachable code here. Is it possible this is an issue with the `[[fallthrough]]` attribute? Would it be possible for you to reproduce the error using [godbolt.org](https://godbolt.org/)? I think we'd all like to better understand what is being hit here.

shushanhf replied:
> I don't really understand what we are missing here. There is no unreachable code here. Is it possible this is an issue with the `[[fallthrough]]` attribute? 

Yes, I think it is the `[[fallthrough]]` attribute as if you add it here it will be built failed on AMD64/ARM64/LoongArch64 linux.

> Would it be possible for you to reproduce the error using [godbolt.org](https://godbolt.org/)? I think we'd all like to better understand what is being hit here.

I will try but I'm not sure I can access this.
But I will rebase some patches on AMD64 linux to find why the `[[fallthrough]]` attribute causes building errors.

AaronRobinsonMSFT replied:
It is possible the solution is to not use the `[[fallthrough]]` attribute on LoongArch64. I don't see this on amd64, so it is either the compiler version being used here or LoongArch64.

shushanhf replied:
> It is possible the solution is to not use the `[[fallthrough]]` attribute on LoongArch64. I don't see this on amd64, so it is either the compiler version being used here or LoongArch64.

AaronRobinsonMSFT replied:
Okay, then this must be a compiler version issue. I can't find arm64 linux with clang-8 on godbolt.org and LoongArch64-linux with clang-13 doesn't seem to repro on gotbolt.org. It also seems like an older version of the compiler than we normally build with. Is it possible for you to move up to at least clang-16?

shushanhf replied:
> Okay, then this must be a compiler version issue. I can't find arm64 linux with clang-8 on godbolt.org and LoongArch64-linux with clang-13 doesn't seem to repro on gotbolt.org. It also seems like an older version of the compiler than we normally build with. Is it possible for you to move up to at least clang-16?

At least on Apr 3 the ARM64/AMD64/LoongArch64 is OK within the same OS env. So I don't think that is for the clang version.

shushanhf replied:
I had found that commit ID `5c4e2a301ec1dd8ef75a3ba98e5be1ff2cea0130` which is #98336  introduced this building error for ARM64/AMD64/LoongArch64 linux.

jkoritzinsky replied:
My guess is that Clang is able to do better control flow analysis now that we're using standard headers (ex the standard assert's implementation in debug is annotated as `noreturn` so Clang knows that the fallthrough will never be hit).

Are you using clang-analyzer?

If this is the cause, I think the correct solution here would be to make the fallthrough release-only.

shushanhf replied:
I didn't try to find the detail reasons.
But I agree with you that what you said makes sense.

Because it's ok for the ARM64/AMD64/LoongArch64 linux if building on release mode.

shushanhf replied:
> My guess is that Clang is able to do better control flow analysis now that we're using standard headers (ex the standard assert's implementation in debug is annotated as `noreturn` so Clang knows that the fallthrough will never be hit).
> 
> Are you using clang-analyzer?
> 
> If this is the cause, I think the correct solution here would be to make the fallthrough release-only.

As the `_ASSERTE(false && "Unknown result value");` is const false that make the debug is annotated as `noreturn` and  Clang knows that the fallthrough will never be hit.
Here I just rewrite the assert which doesn't look like so obvious and to be optimized as const false.

am11 replied:
@shushanhf thanks for the information. Yes, clang-13 is triggering Wimplicit-fallthrough with main, but clang 14 and above work. I'd say it was a bug in clang <= v13, which has been fixed since v14. Since we support older versions, this PR is the right fix. Could you update the comment to specifically suggest something like "because clang versions before 14 had a bug..."? This way when we will re-baseline the minimum requirement, we will have this context.

> 5c4e2a301ec1dd8ef75a3ba98e5be1ff2cea0130 

Yes, previous commit doesn't have the issue. BTW, you can delete `#ifndef FALLTHROUGH` block from `src/coreclr/pal/inc/pal.h` which that commit added, because another commit that day unified the definition c08bd7b666430bbd7bb932281f74c89a7d094204. Please delete the redundant defintion  from pal.h (it's covered by minipal which is already included in PAL and other places).

am11 replied:
@shushanhf thanks for the information. Yes, clang-13 is triggering Wimplicit-fallthrough with main, but clang 14 and above work. I'd say it was a bug in clang <= v13, which has been fixed since v14. Since we support older versions, this PR is the right fix. Could you update the comment to specifically suggest something like "because clang versions before 14 had a bug..."? This way when we will re-baseline the minimum requirement, we will have this context.

> 5c4e2a301ec1dd8ef75a3ba98e5be1ff2cea0130 

Yes, previous commit doesn't have the issue. BTW, you can delete `#ifndef FALLTHROUGH` block from `src/coreclr/pal/inc/pal.h` which that commit added, because another commit that day unified the definition c08bd7b666430bbd7bb932281f74c89a7d094204.

shushanhf replied:
Thanks very much for your professional answer!
I will update it.

jkotas replied:
If it is an issue with specific clang versions, can the workaround be placed under `__clang_major__` ifdef?

am11 replied:
That would be better: `#if defined(__clang__) && (__clang_major__ < 14)`

shushanhf replied:
> If it is an issue with specific clang versions, can the workaround be placed under `__clang_major__` ifdef?

Is it liking this
```diff
+                    _ASSERTE(
+#if defined(__clang_major__) && (__clang_major__ <= 13)
+                            (result == TryInvokeICustomQueryInterfaceResult::FailedToInvoke)
+#else
+                            false
+#endif
+                            && "Unknown result value");
```

am11 replied:
Or maybe wrap the whole line in condition:

```c
// comment here
#if defined(__clang_major__) && (__clang_major__ <= 13)
    _ASSERTE(result == TryInvokeICustomQueryInterfaceResult::FailedToInvoke && "Unknown result value");
#else
    _ASSERTE(false && "Unknown result value");
#endif
```

shushanhf commented:
> Is this on a `Release` or `Debug`/`Checked` build?

On Debug build on the LoongArch64-linux with clang-13 and arm64 linux with clang-8.

shushanhf committed:
Fix building error for `error: fallthrough annotation in unreachable code`
when building on debug mode on loongarch64/arm64/amd64-linux by clang<14.

Also delete `#ifndef FALLTHROUGH` block as #98336 added after #98712.

AaronRobinsonMSFT commented on src/coreclr/interop/comwrappers.cpp:
```diff
@@ -736,7 +736,14 @@ HRESULT ManagedObjectWrapper::QueryInterface(
                     return E_NOINTERFACE;
 
                 default:
+#if defined(__clang__) && (__clang_major__ < 14)
+                    // NOTE: here the `result == *` of assert is just for avoiding the compiling error
+                    // `fallthrough annotation in unreachable code` within debug mode using clang <=13.
+                    // Because clang versions before 14 had a bug here.
+                    _ASSERTE((result == TryInvokeICustomQueryInterfaceResult::FailedToInvoke) && "Unknown result value");
+#else
                     _ASSERTE(false && "Unknown result value");
+#endif
```
```suggestion
_ASSERTE((result == TryInvokeICustomQueryInterfaceResult::FailedToInvoke) && "Unknown result value");
```

Let's just do the thing that should work in all cases rather than the noise.

am11 commented on src/coreclr/interop/comwrappers.cpp:
```diff
@@ -736,7 +736,14 @@ HRESULT ManagedObjectWrapper::QueryInterface(
                     return E_NOINTERFACE;
 
                 default:
+#if defined(__clang__) && (__clang_major__ < 14)
+                    // NOTE: here the `result == *` of assert is just for avoiding the compiling error
+                    // `fallthrough annotation in unreachable code` within debug mode using clang <=13.
+                    // Because clang versions before 14 had a bug here.
+                    _ASSERTE((result == TryInvokeICustomQueryInterfaceResult::FailedToInvoke) && "Unknown result value");
+#else
                     _ASSERTE(false && "Unknown result value");
+#endif
```
nit: `> 13`.

AaronRobinsonMSFT replied:
That works for me.

shushanhf replied:
done

AaronRobinsonMSFT reviewed (approved):
Thank you.
